db/version_set.cc

```diff
@@ -5153,6 +5175,17 @@ Status VersionSet::ProcessManifestWrites(
   assert(pending_manifest_file_number_ == 0);
   auto new_manifest_force =
       new_manifest_on_next_update_.exchange(false, std::memory_order_relaxed);
+
+  // Generate new replication epochs when replication epoch has changed.
+  // `new_manifest_on_next_update_` will only be true when epoch has changed
+  if (new_manifest_force && is_leader) {
+    auto firstMUS = first_writer.edit_list.front()->GetManifestUpdateSequence();
+    auto currentEpoch = current_replication_epoch_.load(std::memory_order_relaxed);
+    ReplicationEpochAddition epochAddition(currentEpoch, firstMUS);
+    first_writer.edit_list.front()->addReplicationEpoch(epochAddition);
+    new_replication_epochs.emplace_back(std::move(epochAddition));
+  }
```

Comment:
Could we have `NewManifestOnNextUpdate` carry the replication epoch and have `new_manifest_on_next_update_` be an int instead of bool? That way you don't need `initial_replication_epoch` in `DBOptions`.

Reply:
Yeah of course, `new_manifest_on_next_update_` needs to be `std::optional<uint64_t>` and optional should be empty when leader epoch hasn't changed. We can model the optional with a pure uint64_t by using the high bit to represent optional state.

We currently use atomic_bool for `new_manifest_on_next_update_` and rely on external properties to ensure thread safety (in particular, the follower will not attempt writing to the manifest until after `NewManifestOnNextUpdate` concludes). Let's document this assumption in `DB::NewManifestOnNextUpdate()`